### PR TITLE
Fix HTML entity rendering in file comments sidebar

### LINKED ISSUE #28243: Quotation marks get html escape in file comments
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. Open the sidebar for a file, select the comments tab, enter a comment containing a double-quote (`"`) character.
2. Look at the completed comments

### Expected behaviour
The entered double-quote character should display when viewing the comment.

### Actual behaviour
The double-quote character is replaced by the sequence `&quot;`
However, the "Activity" tab correctly shows the double-quote character.

### Server configuration

**Operating system:**

Official Docker Apache container.

**Web server:**

Apache

**Database:**

mariadb

**PHP version:**

7.4.21

**Nextcloud version:** (see Nextcloud admin page)

22.0.0

**Updated from an older Nextcloud/ownCloud or fresh install:**

Fresh install

**Where did you install Nextcloud from:**

docker.io/library/nextcloud

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.7.0
  - activity: 2.15.0
  - camerarawpreviews: 0.7.11
  - circles: 22.0.0
  - cloud_federation_api: 1.4.0
  - comments: 1.11.0
  - contacts: 4.0.1
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - dav: 1.18.0
  - facerecognition: 0.8.3
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_antivirus: 3.2.1
  - files_pdfviewer: 2.3.0
  - files_rightclick: 1.1.0
  - files_sharing: 1.13.2
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.11.0
  - firstrunwizard: 2.11.0
  - groupfolders: 9.0.2
  - logreader: 2.7.0
  - lookup_server_connector: 1.9.0
  - maps: 0.1.9
  - nextcloud_announcements: 1.11.0
  - notifications: 2.10.1
  - oauth2: 1.9.0
  - password_policy: 1.12.0
  - photos: 1.4.0
  - previewgenerator: 3.1.1
  - privacy: 1.6.0
  - provisioning_api: 1.11.0
  - recommendations: 1.1.0
  - registration: 1.3.0
  - serverinfo: 1.12.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - spreed: 12.0.1
  - support: 1.5.0
  - survey_client: 1.10.0
  - systemtags: 1.11.0
  - terms_of_service: 1.8.0
  - text: 3.3.0
  - theming: 1.12.0
  - twofactor_backupcodes: 1.10.1
  - updatenotification: 1.11.0
  - user_status: 1.1.1
  - viewer: 1.6.0
  - weather_status: 1.1.0
  - welcome: 0.0.5
  - workflowengine: 2.3.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "gmsmrz.ooguy.com",
            "predator.local"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "22.0.0.11",
        "overwrite.cli.url": "http:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "default_phone_region": "US",
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "ssl",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "app_install_overwrite": [
            "camerarawpreviews",
            "previewgenerator",
            "groupfolders"
        ],
        "maintenance": false
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...

No

**Are you using encryption:** yes/no

No

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

No

### Client configuration
**Browser:**

Chrome

**Operating system:**

Fedora 34

#### Web server error log
<details>
<summary>Web server error log</summary>

```
I don't know how to find this. I'll figure it out if you really need it.
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
This is longer than the scrollback on my terminal. If you really need it, I can try to find a way to get it out of the container and attach it.
```

</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
IRefused to execute script from 'https://predator.local:8080/apps/dashboard/' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
:8080/ocs/v2.php/apps/text/workspace?path=%2F:1 Failed to load resource: the server responded with a status of 404 ()
:8080/ocs/v2.php/apps/text/workspace?path=%2Floose:1 Failed to load resource: the server responded with a status of 404 ()
:8080/ocs/v2.php/apps/spreed/api/v1/file/1131:1 Failed to load resource: the server responded with a status of 404 ()
8Mixed Content: The page at '<URL>' was loaded over HTTPS, but requested an insecure element '<URL>'. This request was automatically upgraded to HTTPS, For more information see <URL>
:8080/ocs/v2.php/apps/spreed/api/v1/file/1131:1 Failed to load resource: the server responded with a status of 404 ()
:8080/ocs/v2.php/apps/user_status/api/v1/statuses/undefined:1 Failed to load resource: the server responded with a status of 404 ()
:8080/ocs/v2.php/apps/spreed/api/v1/file/1131:1 Failed to load resource: the server responded with a status of 404 ()
994util.js:135 OC.Util.relativeModifiedDate is deprecated and will be removed in Nextcloud 21. See @nextcloud/moment
relativeModifiedDate @ util.js:135
DevTools failed to load source map: Could not load content for https://predator.local:8080/js/activity/activity-sidebar.js.map?v=fe49d7c520f13702fb7b: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE
DevTools failed to load source map: Could not parse content for https://predator.local:8080/custom_apps/facerecognition/js/personssidebar.js.map: Unexpected token < in JSON at position 0
DevTools failed to load source map: Could not parse content for https://predator.local:8080/custom_apps/facerecognition/js/vendor/autocomplete.js.map: Unexpected token < in JSON at position 0
254util.js:135 OC.Util.relativeModifiedDate is deprecated and will be removed in Nextcloud 21. See @nextcloud/moment
relativeModifiedDate @ util.js:135
(anonymous) @ init.js:101
each @ jquery.js:354
each @ jquery.js:189
(anonymous) @ init.js:100

```
</details>

### PR DESCRIPTION
**Pull request by Pytal**
Fix #28243

This fix decodes HTML entities contained in comment messages and displaynames.

Some characters are double-encoded, e.g. `>` - `&amp;gt;`, so we decode the strings twice. Full character to encoded version map:

- `>` - `&amp;gt;` double-encoded
- `<` - `&amp;lt;` double-encoded
- `&` - `&amp;` encoded
- `"` - `&quot;` encoded
- `'` - `'` not encoded

### Further work
Update backend XML encoder logic to address inconsistent encoding https://github.com/nextcloud/server/issues/29306

### CODE REVIEW

File: apps/comments/src/components/Comment.vue [hunk 1]
```diff
@@ -180,7 +180,12 @@ export default {
 			if (this.isEmptyMessage) {
 				return ''
 			}
-			return this.renderContent(this.localMessage)
+			// Double-unescape potentially double-escaped localMessage
+			let message = this.localMessage
```
Comment (PVince81) [inline apps/comments/src/components/Comment.vue:184]: would it be more suitable to parse it at an earlier time, close to the "parseXML" call from when messages are retrieved ? this way it stays consistent everywhere, not only at render time

Comment (Pytal) [reply apps/comments/src/components/Comment.vue]: Yep, updated and now displaynames with HTML entities also render correctly :)

File: apps/comments/src/services/GetComments.js [hunk 2]
```diff
@@ -74,6 +74,22 @@ function processMultistatus(result, isDetailed = false) {
 		const {
 			propstat: { prop: props },
 		} = item
-		return prepareFileFromProps(props, props.id.toString(), isDetailed)
+		// Decode HTML entities
+		const decodedProps = {
+			...props,
+			// Decode twice to handle potentially double-encoded entities
```
Comment (PVince81) [inline apps/comments/src/services/GetComments.js:80]: it feels a bit risky to always double decode when we are not sure if this is necessary

did you find the exact scenarios where double encoding is happening ?
in the original report there was only single encoding on read

Comment (Pytal) [reply apps/comments/src/services/GetComments.js:80]: Yes, PR description updated with my findings :), the double decode is just to handle the additional cases

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by Pytal** [comment]
@tobiasKaminsky do clients render these correctly?

**Comment by tobiasKaminsky** [comment]
Android: 
- input on Android: works
- input on Server: we just show activity feed per file, so we show what server renders -> Android not affected

Desktop:
- does not have comment function
- only shows comments via activity stream

iOS:
- could not test this PR
- 22.0.0 is working correctly, so if you do not change anything on backend logic, but only how comments are displayed, it should be fine.

**Review by CarlSchwan — APPROVED** [review]
Looks good to me, I will rebase it

**Pull request merged**